### ISSUE TITLE
OpenAPI handler fails on duplicated path

### ISSUE THREAD

**Issue by natherz97** [kubernetes/kubernetes#122668]
### What happened?

When installing the metrics-server to implement the metrics.k8s.io API and the prometheus-adapter to implement the custom.metrics.k8s.io API, it is possible to make the OpenAPI handler fail to initialize with this error log in apiserver: https://github.com/kubernetes/kube-openapi/blob/master/pkg/aggregator/aggregator.go#L343
```
handler.go:159] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/metrics.k8s.io/v1beta1/pods
handler.go:159] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/metrics.k8s.io/v1beta1/namespaces/{namespace}/pods
handler.go:159] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/metrics.k8s.io/v1beta1/nodes
handler.go:159] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/metrics.k8s.io/v1beta1/nodes/{name}
handler.go:159] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/metrics.k8s.io/v1beta1/nodes
```
This will result in kubectl discovery calls to /openapi/v2 to fail with a 503:
```
{
  "kind": "Event",
  "apiVersion": "audit.k8s.io/v1",
  "level": "Metadata",
  "auditID": "863d36c0-3797-4f19-bc74-df571a671a00",
  "stage": "ResponseComplete",
  "requestURI": "/openapi/v2?timeout=31s",
  "verb": "get",
...
  "userAgent": "kubectl/v1.28.2 (linux/amd64) kubernetes/cd9e211",
  "responseStatus": {
    "metadata": {},
    "code": 503
  },
  "requestReceivedTimestamp": "2023-12-19T02:50:12.882602Z",
  "stageTimestamp": "2023-12-19T02:50:12.913981Z",
  "annotations": {
    "authorization.k8s.io/decision": "allow",
    "authorization.k8s.io/reason": ""
  }
}
```

### What did you expect to happen?

The OpenAPI handler should not fail to initialize if both metrics.k8s.io and custom.metrics.k8s.io are serving the same path for different API groups/versions.

### How can we reproduce it (as minimally and precisely as possible)?

I can provide exact installation steps if needed, however, a high-level repro is:
- install metrics-server to serve v1beta1.metrics.k8s.io  
- install prometheus-adapter to serve v1beta1.custom.metrics.k8s.io
- ensure resourceRules is populated in the prometheus-adapter config so that resource metrics are also served by v1beta1.custom.metrics.k8s.io: https://github.com/kubernetes-sigs/prometheus-adapter/blob/7a421bb91e9ea52be4e26db22392c36c8af38bc8/deploy/manifests/custom-metrics-config-map.yaml#L73-L98
- restarting apiserver will result in the OpenAPI failure and 503 responses to /openapi/v2 path

### Kubernetes version

<details>

```console
$ kubectl version
Server Version: version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.4-eks-8cb36c9", GitCommit:"b4ddd686b005f445825bd0afd5dba24d4414d4cf", GitTreeState:"clean", BuildDate:"2023-11-22T21:52:14Z", GoVersion:"go1.20.11", Compiler:"gc", Platform:"linux/amd64"}
```
</details>

### Cloud provider

<details>
AWS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by natherz97** [comment]
Please let me know if I should reopen the issue in https://github.com/kubernetes/kube-openapi.

**Comment by BenTheElder** [comment]
/sig instrumentation
[metrics apis]

**Comment by cici37** [comment]
@Jefftree Would you mind taking a look when have time? Thank you^^

/triage accepted

**Comment by Jefftree** [comment]
I have a question. `custom.metrics.k8s.io` should be serving under `custom.metrics.k8s.io/v1beta1`. Why does an apiservice serving another gvk register paths under `metrics.k8s.io/v1beta1`?

cc @deads2k, we've made the assumption that an APIService (such as metrics-server or prometheus-adapter) serves only gvks under its own registered group version. I'll do some more investigation but just wondering if prometheus doesn't follow that assumption?

**Comment by natherz97** [comment]
@Jefftree the prometheus-adapter can also be configured to serve metrics.k8s.io and replace the metrics-server by enabling these metrics, https://github.com/kubernetes-sigs/prometheus-adapter/blob/7a421bb91e9ea52be4e26db22392c36c8af38bc8/deploy/manifests/custom-metrics-config-map.yaml#L73-L98. 

However, if the prometheus-adapter fronts custom.metrics.k8s.io and metrics-server fronts metrics.k8s.io, it does seem like the correct behavior would be for prometheus-adapter to serve /apis/custom.metrics.k8s.io/v1beta1/nodes until an operator updates the metrics.k8s.io APIService to reference prometheus-adapter rather than metrics-server. We're seeing this issue occur for upgrades between 1.27 and 1.28 so it does look this path duplication did not cause the OpenAPI handler to fail on 1.27.

**Comment by dobesv** [comment]
I just upgraded from 1.27 to 1.28 and running into this problem. Some kubectl apply operations fail, and ArgoCD cannot load its custom resources.  Any suggested workaround?

Would removing the `resourceRules` from the prometheus configuration resolve it?  Or perhaps disable metrics-server?

**Comment by Jefftree** [comment]
One other question is if this is reproducible by just applying a set of yaml/installations or does it require a cluster upgrade to trigger? Thanks!

**Comment by chaunceyctx97** [comment]
@Jefftree 
hit same problem
```
1. install metrics-server to serve v1beta1.metrics.k8s.io
2. install prometheus-adapter to serve v1beta1.custom.metrics.k8s.io
3. restarting 1.28 kube-apiserver will result in this failure
```

and we cannot install any other new helm plugins. maybe we need to move forward

**Comment by natherz97** [comment]
@Jefftree Here are verbose repro instructions. An apiserver restart on 1.28 will induce the issue.
```
# install metrics-server
$ kubectl apply -f https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/high-availability-1.21+.yaml

# install prometheus
$ kubectl create namespace prometheus
$ helm upgrade -i prometheus prometheus-community/prometheus \
    --namespace prometheus

# install prometheus-adapter
$ helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
$ helm repo update
$ helm install my-release prometheus-community/prometheus-adapter

# edit prometheus-adapter to point to prometheus installation
$ kubectl edit deployment my-release-prometheus-adapter
...
- --prometheus-url=http://prometheus-server.prometheus.svc:80

# enable prometheus-adapter to serve metrics.k8s.io. Edit the configmap and add these lines, https://github.com/kubernetes-sigs/prometheus-adapter/blob/7a421bb91e9ea52be4e26db22392c36c8af38bc8/deploy/manifests/custom-metrics-config-map.yaml#L73-L98
$ kubectl edit configmap my-release-prometheus-adapter

# prometheus-adapter-resource-reader clusterrole may need updated for additional permissions
$ kubectl edit clusterrole prometheus-adapter-resource-reader

```

**Comment by Jefftree** [comment]
Hello. I'm still having some trouble reproducing although I'm not 100% sure if the configuration is correct. I'm trying to repro via kind.

afaik installing via helm has prometheus adapter serving `custom.metrics.k8s.io/v1beta1`. [The latest version of the apiservice](https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/deploy/manifests/api-service.yaml) seems to suggest that prometheus adapter is serving under `metrics.k8s.io/v1beta1`. Applying the update to the configmap per the instructions didn't change the apiservice on my end. Not sure if this is my misconfiguration or intended.

I've also tested via integration test that having two openapis serving the same path doesn't cause errors and one will get filtered out.

A simple workaround is probably to delete the metrics-server if it is already superseded by the prometheus-adapter.

**Comment by dobesv** [comment]
The `metrics.k8s.io/v1beta1` is not set up in an `APIService` resource for us, we only created an APIService with `custom.metrics.k8s.io/v1beta1`.

When I go fetch the openAPI schema served up by the endpoint, it has schema information for `metrics.k8s.io/v1beta1` in it if that is enabled using `resourceRules`.

If I remove `resourceRules` then the schema fetched from the server doesn't have `metrics.k8s.io/v1beta1` stuff in it, and the error is resolved.

We have been keeping `metrics-server` around because I think something uses it directly, or used to.  Maybe `kubernetes-dashboard`? That could have changed 🤔 

But in any case, this problem is quite severe when it happens so it would be nice to fix it as there may be others who do the same update and have to do some deep diving into the logs and all kinds of crazy diagnostics and guesswork to figure it out.

**Comment by Jefftree** [comment]
How does the metrics.k8s.io/v1beta1 get reached without an APIService? The resources would be unavailable for the apiserver to query and it'll be through something else?

I see, adding resources for me doesn't populate the metrics.k8s.io/v1beta1 OpenAPI. I wonder if this could be cloud provider specific.

I'm totally supportive of fixing, it's just quite difficult to reproduce on my part 😅.

**Comment by dobesv** [comment]
Hmm good question.  In our staging cluster I tried bringing back the `resourceRules` in the custom metrics server configuration and restarting all the masters, and the problem did not come back, everything seems OK now. So it maybe isn't that easy to reproduce 🤔

**Comment by natherz97** [comment]
@Jefftree After updating the configmap, restarting the prometheus-adapter pod to pickup the changes and restarting the apiserver are both required to induce the issue. I have been able to deterministically reproduce this following those steps, please follow up here if you are still having issues.

**Comment by Jefftree** [comment]
@natherz97 That's good to hear. I'll need some more help with reproducing it. I've been able to verify that restarting the prometheus adapter pod picks up the configmap changes but restarting the apiserver hasn't caused any crashes for me. I've tested this both on the latest master branch and 1.28.

**Comment by shyamjvs** [comment]
While I'm yet to fully understand these changes, the 2 PRs below (both of which went into 1.28) seem likely candidates for a regression here:

- https://github.com/kubernetes/kubernetes/pull/118212 - introduced lazy loading in openAPI v2 handler to load/compute the aggregator spec when requested by client (cc @Jefftree)
- https://github.com/kubernetes/kubernetes/pull/118204 - switched the spec merging function from `MergeSpecsIgnorePathConflict` to `MergeSpecsIgnorePathConflictRenamingDefinitionsAndParameters` (cc @sttts)

@Jefftree @sttts any clue if one of the above changes could be causing mischief? 

This change that went into 1.28 seems like the most likely cause - https://github.com/kubernetes/kubernetes/pull/118212

**Comment by shyamjvs** [comment]
Also from @natherz97 example request at the top, the impact seems to be to the v2 API:

```
  "requestURI": "/openapi/v2?timeout=31s",
  "verb": "get",
```

Nathan - when you get a chance, could you also confirm if v1 API doesn't run into this?

**Comment by mengqiy** [comment]
@shyamjvs OpenAPI only supports v2 and v3.

**Comment by Chaunceyctx** [comment]
@Jefftree @sttts hi, any clue about this issue?

**Comment by chaunceyctx97** [comment]
@Jefftree @sttts hi, any clue about this issue?

**Comment by Jefftree** [comment]
It's possible but I haven't been able to reproduce it. Can I ask if this is happening on cloud providers outside AWS?

**Comment by natherz97** [comment]
@Jefftree Here are the apiservices and relevant pod objects with a working repro from these steps https://github.com/kubernetes/kubernetes/issues/122668#issuecomment-1917622117:
```
% kubectl version
Server Version: version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.6-eks-508b6b3", GitCommit:"25a726351cee8ee6facce01af4214605e089d5da", GitTreeState:"clean", BuildDate:"2024-01-29T20:58:56Z", GoVersion:"go1.20.13", Compiler:"gc", Platform:"linux/amd64"}

% kubectl get apiservices | grep metrics.k8s.io
v1beta1.custom.metrics.k8s.io          default/my-release-prometheus-adapter   True        2d22h
v1beta1.metrics.k8s.io                 kube-system/metrics-server              True        2d22h

% kubectl get pods -A
NAMESPACE     NAME                                                 READY   STATUS    RESTARTS   AGE
default       my-release-prometheus-adapter-6475854d97-5snjm       1/1     Running   0          2d22h
kube-system   metrics-server-7c4f45f94b-9dl6v                      1/1     Running   0          2d22h
kube-system   metrics-server-7c4f45f94b-mmfm7                      1/1     Running   0          2d22h
prometheus    prometheus-alertmanager-0                            0/1     Pending   0          2d22h
prometheus    prometheus-kube-state-metrics-6b464f5b88-f74tr       1/1     Running   0          2d22h
prometheus    prometheus-prometheus-node-exporter-8n6xd            1/1     Running   0          2d22h
prometheus    prometheus-prometheus-node-exporter-vn6jc            1/1     Running   0          2d22h
prometheus    prometheus-prometheus-pushgateway-7857c44f49-9g6z4   1/1     Running   0          2d22h
prometheus    prometheus-server-75dc84d779-8zwxk                   2/2     Running   0          2d22h
```
If it's useful, I can provide yamls for these objects in addition to the settings applied in the prometheus-adapter configmap.

**Comment by SuQiucheng** [comment]
> It's possible but I haven't been able to reproduce it. Can I ask if this is happening on cloud providers outside AWS?

I meet this problem

**Comment by chaunceyctx97** [comment]
> It's possible but I haven't been able to reproduce it. Can I ask if this is happening on cloud providers outside AWS?

@Jefftree ACK in Alibaba Cloud also meets this problem

**Comment by uwefreidank** [comment]
We were able to fix the v2 availability by `kubectl delete apiservices v1beta1.custom.metrics.k8s.io` and re-installing the kube-prometheus helm chart.
Unfortunately, after restarting the k8s api server, v2 availability is broken again.

**Comment by Jefftree** [comment]
@natherz97 yes, I'm running on kind. 

```
kubernetes ❯ k get pods -A
NAMESPACE            NAME                                                 READY   STATUS             RESTARTS        AGE
...
default              my-release-prometheus-adapter-6b74748d98-xhrrw       1/1     Running            0               5m34s
kube-system          metrics-server-66f57d964b-2zcq4                      1/1     Running            0               2m3s
kube-system          metrics-server-66f57d964b-5tl5z                      1/1     Running            0               2m3s
prometheus           prometheus-alertmanager-0                            1/1     Running            0               10m
prometheus           prometheus-kube-state-metrics-745b475957-7k2f9       1/1     Running            0               10m
prometheus           prometheus-prometheus-node-exporter-8szsq            1/1     Running            0               10m
prometheus           prometheus-prometheus-node-exporter-tjhkx            1/1     Running            0               10m
prometheus           prometheus-prometheus-node-exporter-wrpcn            1/1     Running            0               10m
prometheus           prometheus-prometheus-pushgateway-6ccd698d79-5sfm6   1/1     Running            0               10m
prometheus           prometheus-server-5c99dfc547-ldwc5                   2/2     Running            0               10m
kubernetes ❯ k get apiservices | grep metrics.k8s.io   
v1beta1.custom.metrics.k8s.io     default/my-release-prometheus-adapter   True        9m7s
v1beta1.metrics.k8s.io            kube-system/metrics-server              True        9m28s
```

I'll do some more digging, assuming it's related to those PRs linked by @shyamjvs but still haven't pinpointed anything.

**Comment by hakuna-matatah** [comment]
> - --prometheus-url=http://prometheus-server.prometheus.svc:80

@Jefftree I tried to reproduce on my local setup with `kind` based on steps [here](https://github.com/kubernetes/kubernetes/issues/122668#issuecomment-1917622117) + `apiserver restart`, but I could still get a successful response  when I executed `kubectl get --raw /openapi/v2.

Then I reverted one of the step mentioned in [here](https://github.com/kubernetes/kubernetes/issues/122668#issuecomment-1917622117) to `- --prometheus-url=http://prometheus.default.svc:9090` and waited until the new `my-release-prometheus-adapter` pod came alive and  then did a apiserver restart and I was able to reproduce the same thing on `kind cluster` as well  i.e; I get the 503 response when I execute following command after the fact. 


```
dev-dsk-hakuna-2c-0fa5574b % kubectl get --raw /openapi/v2     --v=9
I0228 05:21:10.106706  144733 loader.go:395] Config loaded from file:  /home/hakuna/.kube/config
I0228 05:21:10.107467  144733 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.29.2 (linux/amd64) kubernetes/4b8e819" 'https://127.0.0.1:43171/openapi/v2'
I0228 05:21:10.107789  144733 round_trippers.go:510] HTTP Trace: Dial to tcp:127.0.0.1:43171 succeed
I0228 05:21:10.135567  144733 round_trippers.go:553] GET https://127.0.0.1:43171/openapi/v2 503 Service Unavailable in 28 milliseconds
I0228 05:21:10.135593  144733 round_trippers.go:570] HTTP Statistics: DNSLookup 0 ms Dial 0 ms TLSHandshake 5 ms ServerProcessing 22 ms Duration 28 ms
I0228 05:21:10.135615  144733 round_trippers.go:577] Response Headers:
I0228 05:21:10.135632  144733 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: bcaa6da0-0e9d-4093-84eb-657d80be2b85
I0228 05:21:10.135649  144733 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 97870445-4efc-44b1-8ecb-e0fbcb170a0e
I0228 05:21:10.135657  144733 round_trippers.go:580]     Content-Length: 0
I0228 05:21:10.135667  144733 round_trippers.go:580]     Date: Wed, 28 Feb 2024 05:21:10 GMT
I0228 05:21:10.135681  144733 round_trippers.go:580]     Audit-Id: 08d5275e-357e-43e9-9248-04536b3bd006
I0228 05:21:10.135689  144733 round_trippers.go:580]     Cache-Control: no-cache, private
I0228 05:21:10.135702  144733 round_trippers.go:580]     Vary: Accept-Encoding
I0228 05:21:10.135719  144733 round_trippers.go:580]     Vary: Accept
I0228 05:21:10.135756  144733 request.go:1212] Response Body: 
I0228 05:21:10.135905  144733 helpers.go:246] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "the server is currently unable to handle the request",
  "reason": "ServiceUnavailable",
  "details": {
    "causes": [
      {
        "reason": "UnexpectedServerResponse"
      }
    ]
  },
  "code": 503
}]
Error from server (ServiceUnavailable): the server is currently unable to handle the request

```

After this point, even if i change it back to `- --prometheus-url=http://prometheus-server.prometheus.svc:80` and restart the apiserver ,  the error is persistent. Let me know if this works for you to reproduce ?

**Comment by Jefftree** [comment]
Thanks, alright I found a way to reproduce it consistently without restarting the apiserver. Still investigating the root cause.

**Comment by Jefftree** [comment]
Quick update: This issue was a combination of bugs both in k/k and prometheus adapter. I opened https://github.com/kubernetes/kubernetes/pull/123570 to mitigate on k/k side which should fix the issue. 

The issue was difficult to reproduce because it requires a very specific ordering of OpenAPI merging to trigger, which is why I suspect many people were able to reproduce after some restarts to kube-apiserver. 

Prometheus and metrics server serve differing OpenAPIs for the `/metrics.k8s.io/v1beta1` endpoint. Prometheus also doesn't seem to be registering discovery's OpenAPI properly (missing OpenAPI for `/metrics.k8s.io/` but having the OpenAPI for `/metrics.k8s.io/v1beta1/{endpoints}` whereas it is all present in metrics server. This inconsistency in how prometheus and metrics server handle the OpenAPI surfaced the merging issue.

**Comment by chaunceyctx97** [comment]
> Quick update: This issue was a combination of bugs both in k/k and prometheus adapter. I opened #123570 to mitigate on k/k side which should fix the issue.
> 
> The issue was difficult to reproduce because it requires a very specific ordering of OpenAPI merging to trigger, which is why I suspect many people were able to reproduce after some restarts to kube-apiserver.
> 
> Prometheus and metrics server serve differing OpenAPIs for the `/metrics.k8s.io/v1beta1` endpoint. Prometheus also doesn't seem to be registering discovery's OpenAPI properly (missing OpenAPI for `/metrics.k8s.io/` but having the OpenAPI for `/metrics.k8s.io/v1beta1/{endpoints}` whereas it is all present in metrics server. This inconsistency in how prometheus and metrics server handle the OpenAPI surfaced the merging issue.

@Jefftree a little wired question: why do we need to restart kube-apiserver to reproduce this problem?

**Comment by sushanth0910** [comment]
Tried to reproduce the issue after cherry picking the fix https://github.com/kubernetes/kubernetes/pull/123570 and didn't run into below error.
`handler.go:159] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/metrics.k8s.io/v1beta1/pods`

Fix is working as expected.

**Comment by olvesh** [comment]
> I just upgraded from 1.27 to 1.28 and running into this problem. Some kubectl apply operations fail, and ArgoCD cannot load its custom resources. Any suggested workaround?
> 
> Would removing the `resourceRules` from the prometheus configuration resolve it? Or perhaps disable metrics-server?


It can look like restarting prometheus-adapter "fixes" the issue? I am also curious about @dobesv suggestion of removing the `resourceRules` section from the config map works as a temporary workaround until released upstream (we use `externalMetrics` and cannot delete the apiregistration)?

As I understand the fix for this issue is currently in 1.30.0.beta.0
```
❯ git tag --contains 9a404b590b1e451e8349eb9730d93167c9650fed
v1.30.0-beta.0
v1.31.0-alpha.0
```

**Comment by raffis** [comment]
@yogeek Any chance you are using an older version of kubectl?

**Comment by raffis** [comment]
> @Jefftree Sorry to comment on a closed issue but I just stumbled into a similar issue in my 1.28 cluster which in theory should benefit from the backported fix

We have the same issue here, actually also triggered by installing the kyverno reports-server with both apiservices (different groups served by the same api-server).
Kubernetes 1.29.9 though. The 1.29 branch contains this fix but we still experience problems.
Client write request experience random `the server is currently unable to handle the request` responses while on the kube-apiserver we see 
```
handler.go:160] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/reports.kyverno.io/v1/namespaces/{namespace}/ephemeralreports/{name}
```

**Comment by Jefftree** [comment]
Can this be reproduced without an upgrade or is that the main path of triggering this?

**Comment by raffis** [comment]
> Can this be reproduced without an upgrade or is that the main path of triggering this?

I tried yesterday to reproduce it with a kind cluster from scratch, install kyverno-reports server, kyverno crds and was unable to reproduce it even with the exact same api-server image we have on our kops aws cluster.

**Comment by dims** [comment]
For those who land up here...

There are two PRs linked to [issue 123570](https://github.com/kubernetes/kubernetes/pull/123570) which fixes this issue:

- https://github.com/kubernetes/kubernetes/pull/123570
- https://github.com/kubernetes/kubernetes/pull/123770

Commits from both PRs ended up all the way back to 1.28 as well in a single PR:
https://github.com/kubernetes/kubernetes/pull/123625/commits

The changes landed in 1.28 branch back in March.

**Comment by kumarkeshav021** [comment]
> > @Jefftree Sorry to comment on a closed issue but I just stumbled into a similar issue in my 1.28 cluster which in theory should benefit from the backported fix
> 
> We have the same issue here, actually also triggered by installing the kyverno reports-server with both apiservices (different groups served by the same api-server). Kubernetes 1.29.9 though. The 1.29 branch contains this fix but we still experience problems. Client write request experience random `the server is currently unable to handle the request` responses while on the kube-apiserver we see
> 
> ```
> handler.go:160] Error in OpenAPI handler: failed to build merge specs: unable to merge: duplicated path /apis/reports.kyverno.io/v1/namespaces/{namespace}/ephemeralreports/{name}
> ```
> 
> kubectl get --raw /openapi/v2 fails all the time which I assume is invoked occasionally by kubectl write requests? Thats why the error only happens randomly I assume.

I am getting this on k8s 1.30 with deployment of kyverno reports-server . I had to restart the APIService to reproduce the error

**Comment by oarribas** [comment]
@kumarkeshav021 , I have seen the bug reported by you to Kyverno [1]. In that bug, you mentioned the issue started when you upgraded your Kubernetes. Could you try what was suggested above to reproduce this in a new Kubernetes installation, as looks like it was not reproducible as per [2]?

>>  Can this be reproduced without an upgrade or is that the main path of triggering this?
>
>I tried yesterday to reproduce it with a kind cluster from scratch, install kyverno-reports server, kyverno crds and was unable to reproduce it even with the exact same api-server image we have on our kops aws cluster.


[1] https://github.com/kyverno/reports-server/issues/243
[2] https://github.com/kubernetes/kubernetes/issues/122668#issuecomment-2534225226

**Comment by kumarkeshav021** [comment]
@oarribas Yes i did try to install on fresh k8s , it works but once kube-apiserver restarted i see the same error
I tested on 1.28, 1.29 and 1.30 . Got the same result on all

**Comment by raffis** [comment]
For the record I resolved the problem (at least for now it is working) by removing the 4 reports related kyverno crds. Theoretically they shouldn't be required as the apis are implemented by the aggregated server.
I assume somehow the crds presence can clash with the aggregated apiserver although I was unable to reproduce it in a separate env. Might also be some form of a race condition.

However the problem is that the kyverno controllers including reports-controller check at startup whether those crds exist. So I acually had to readd them afterwards. But apparantly after I readded them I did not run into the same problem anymore.
`kubectl  get --raw /openapi/v2` is working ever since I did that. I raised this here https://github.com/kyverno/kyverno/pull/11867
